consensus/serenity/serenity.go

```diff
@@ -87,54 +95,62 @@ func (s *Serenity) VerifyUncles(chain consensus.ChainReader, header *types.Heade
 
 // Prepare makes sure difficulty and nonce are correct
 func (s *Serenity) Prepare(chain consensus.ChainHeaderReader, header *types.Header) error {
+	reached, err := IsTTDReached(chain, header.ParentHash, header.Number.Uint64()-1)
+	if err != nil {
+		return err
+	}
+	if !reached {
+		return s.eth1Engine.Prepare(chain, header)
+	}
 	header.Difficulty = SerenityDifficulty
-	header.Nonce = types.BlockNonce{}
+	header.Nonce = SerenityNonce
 	return nil
 }
 
+func (s *Serenity) Finalize(config *params.ChainConfig, header *types.Header, state *state.IntraBlockState,
+	txs []types.Transaction, uncles []*types.Header, r types.Receipts, e consensus.EpochReader, chain consensus.ChainHeaderReader,
+	syscall consensus.SystemCall) (systemTxs []types.Transaction, usedGas uint64, err error) {
+	if !IsPoSHeader(header) {
+		return s.eth1Engine.Finalize(config, header, state, txs, uncles, r, e, chain, syscall)
+	}
+	return nil, 0, nil
+}
+
 func (s *Serenity) FinalizeAndAssemble(config *params.ChainConfig, header *types.Header,
 	state *state.IntraBlockState, txs []types.Transaction, uncles []*types.Header,
 	receipts types.Receipts, e consensus.EpochReader, chain consensus.ChainHeaderReader,
-	syscall consensus.SystemCall, call consensus.Call) (*types.Block, error) {
-	return types.NewBlock(header, txs, uncles, receipts), nil
+	syscall consensus.SystemCall, call consensus.Call) (*types.Block, []*types.Receipt, error) {
+	if !IsPoSHeader(header) {
+		return s.eth1Engine.FinalizeAndAssemble(config, header, state, txs, uncles, receipts, e, chain, syscall, call)
+	}
+	return types.NewBlock(header, txs, uncles, receipts), receipts, nil
 }
 
 func (s *Serenity) SealHash(header *types.Header) (hash common.Hash) {
-	hasher := sha3.NewLegacyKeccak256()
-
-	enc := []interface{}{
-		header.ParentHash,
-		header.UncleHash,
-		header.Coinbase,
-		header.Root,
-		header.TxHash,
-		header.ReceiptHash,
-		header.Bloom,
-		header.Difficulty,
-		header.Number,
-		header.GasLimit,
-		header.GasUsed,
-		header.Time,
-		header.Extra,
-	}
-	if header.Eip1559 {
-		enc = append(enc, header.BaseFee)
-	}
-	rlp.Encode(hasher, enc)
-	hasher.Sum(hash[:0])
-	return hash
+	return s.eth1Engine.SealHash(header)
 }
 
 func (s *Serenity) CalcDifficulty(chain consensus.ChainHeaderReader, time, parentTime uint64, parentDifficulty *big.Int, parentNumber uint64, parentHash, parentUncleHash common.Hash, parentSeal []rlp.RawValue) *big.Int {
+	reached, err := IsTTDReached(chain, parentHash, parentNumber)
```

Comment:
Again, we need to make sure that if we got a POSHeader before, we do not throw an error

Reply:
I'm going to make a change in a separate PR to write total difficulty for all blocks, incl. PoS ones. I don't want to do it in this PR in order to keep it small.